libcache/cache.c

```diff
@@ -0,0 +1,650 @@
+
+	cache->sets = calloc(cache->numSets, sizeof(cacheset_t));
+
+	if (cache->sets == NULL) {
+		free(cache);
+		return NULL;
+	}
+
+	cache->offBitsNum = LOG2(cache->lineSize);
+	cache->setBitsNum = LOG2(cache->numSets);
+	cache->tagBitsNum = LIBCACHE_ADDR_WIDTH - cache->setBitsNum - cache->offBitsNum;
+
+	cache->tagMask = cache_genMask(cache->tagBitsNum);
+	cache->setMask = cache_genMask(cache->setBitsNum);
+	cache->offMask = cache_genMask(cache->offBitsNum);
+
+	cache->readCb = readCb;
+	cache->writeCb = writeCb;
+
+	return cache;
+}
+
+
+int cache_deinit(cachectx_t *cache)
+{
+	int i, j;
+
+	for (i = 0; i < cache->numSets; ++i) {
+		for (j = 0; j < LIBCACHE_NUM_WAYS; ++j) {
+			if (IS_VALID(cache->sets[i].lines[j].flags)) {
+				free(cache->sets[i].lines[j].data);
+				CLEAR_VALID(cache->sets[i].lines[j].flags);
+			}
+		}
+	}
+	free(cache->sets);
+
+	free(cache);
+
+	return 0;
+}
+
+
+static int cache_compareTags(const void *lhs, const void *rhs)
+{
+	int ret = 0;
+
+	cacheline_t **lhsLine = (cacheline_t **)lhs;
+	cacheline_t **rhsLine = (cacheline_t **)rhs;
+
+	if (*lhsLine == NULL && *rhsLine == NULL) {
+		ret = 0;
+	}
+	else if (*lhsLine == NULL && *rhsLine != NULL) {
+		ret = -1;
+	}
+	else if (*lhsLine != NULL && *rhsLine == NULL) {
+		ret = 1;
+	}
+	else {
+		if ((*lhsLine)->tag < (*rhsLine)->tag) {
+			ret = -1;
+		}
+		else if ((*lhsLine)->tag == (*rhsLine)->tag) {
+			ret = 0;
+		}
+		else {
+			ret = 1;
+		}
+	}
+
+	return ret;
+}
+
+
+static ssize_t cache_flushLine(cache_writeCb_t writeCb, cacheline_t *linePtr, const uint64_t addr, size_t count)
+{
+	ssize_t written = 0;
+
+	if (linePtr != NULL && IS_VALID(linePtr->flags) && IS_DIRTY(linePtr->flags)) {
+		written = (*writeCb)(addr, linePtr->data, sizeof(*linePtr->data), count);
+		CLEAR_DIRTY(linePtr->flags);
+		if (written != count) {
+			written = -1;
+		}
+	}
+
+	return written;
+}
+
+
+static ssize_t cache_executePolicy(cache_writeCb_t writeCb, cacheline_t *linePtr, const uint64_t addr, size_t count, int policy)
+{
+	ssize_t written = 0;
+
+	switch (policy) {
+		case 0:
+			/* write-back */
+			SET_DIRTY(linePtr->flags);
+			written = count;
+			break;
+		case 1:
+			/* write-through */
+			written = (*writeCb)(addr, linePtr->data, sizeof(*linePtr->data), count);
+			break;
+		default:
+			break;
+	}
+
+	return written;
+}
+
+
+static ssize_t cache_writeToSet(cache_writeCb_t writeCb, cacheset_t *set, cacheline_t *line, const uint64_t offset, size_t count, int policy)
+{
+	int i;
+	ssize_t written = 0;
+	cacheline_t *temp = NULL;
+
+	if (set->count < LIBCACHE_NUM_WAYS) {
+		set->count++;
+
+		for (i = 0; i < set->count; ++i) {
+			if (!IS_VALID(set->lines[i].flags)) {
+				set->lines[i] = *line;
+				break;
+			}
+		}
+
+		temp = &(set->lines[i]);
+
+		for (i = 0; i < set->count; ++i) {
+			if (set->tags[i] == NULL) {
+				set->tags[i] = temp;
+				break;
+			}
+		}
+
+		LIST_ADD(&set->timestamps, temp);
+	}
+	else {
+		temp = set->timestamps;
```

Comment:
isn't set->timestamps the most recently accessed?
```suggestion
		temp = set->timestamps->prev;
```

Reply:
Nope, it's ok